### PR TITLE
loading plugin counts the number of outstanding effects per effect, module, and global

### PR DESCRIPTION
**Pull request by crazy4groovy**
I was having problems with the loading module resolving a pending effect too early, because out of a batch of dispatched effects, the first completed one set the loading state to `false`.

This implements an integer counter instead of a boolean value.

Also cleaned up the plugin tests slightly.

### CODE REVIEW

File: plugins/loading/test/loading.test.js [hunk 1]
```diff
@@ -1,201 +1,231 @@
...
+  test('loading.global should be 2 for two dispatched effects', () => {
     const store = init({
       models: { count },
       plugins: [loadingPlugin()]
     })
+
+    dispatch.count.timeout()
     dispatch.count.timeout()
-    expect(store.getState().loading.global).toBe(true)
+    expect(store.getState().loading.global).toBe(2)
```
Comment (crazy4groovy) [inline plugins/loading/test/loading.test.js]: this is the gist of the new functionality

File: package.json [hunk 2]
```diff
@@ -39,7 +39,8 @@
     "test": "jest"
   },
   "dependencies": {
-    "redux": "^3.7.2"
+    "redux": "^3.7.2",
+    "redux-persist": "^5.9.1"
```
[the remark this answers is not in the document]
Comment (ShMcK) [reply package.json]: I noticed that in your PR. Tests "should" pass with 'redux-persist' as a dev dependency.

File: plugins/loading/test/loading.test.js [hunk 3]
```diff
@@ -1,201 +1,231 @@
...
-    const loadingPlugin = require('../src').default
+  test('should change the loading.effects to 2', () => {
+    const store = init({
+      models: { count },
+      plugins: [loadingPlugin()]
+    })
+
+    dispatch.count.timeout()
+    dispatch.count.timeout()
+    expect(store.getState().loading.effects.count.timeout).toBe(2)
```
Comment (crazy4groovy) [inline plugins/loading/test/loading.test.js]: this is the gist of the new functionality

File: plugins/loading/test/loading.test.js [hunk 4]
```diff
@@ -1,201 +1,231 @@
...
+    const store = init({
+      models: { count },
+      plugins: [loadingPlugin()]
+    })
+
+    dispatch.count.timeout1()
+    dispatch.count.timeout2()
+    expect(store.getState().loading.effects.count.timeout1).toBe(1)
+    expect(store.getState().loading.effects.count.timeout2).toBe(1)
+    expect(store.getState().loading.global).toBe(2)
```
Comment (crazy4groovy) [inline plugins/loading/test/loading.test.js]: making sure global increments for every dispatched effect

File: plugins/loading/src/index.ts [hunk 5]
```diff
@@ -55,45 +59,49 @@ const loadingPlugin = (config: LoadingConfig = {}): any => {
...
+          const origEffect = dispatch[name][action]
+
+          // create function with pre & post loading calls
+          const effectWrapper = async (...props) => {
+            try {
+              dispatch.loading.show({ name, action })
+              await origEffect(...props)
+              // waits for dispatch function to finish before calling "hide"
+            } finally {
+              dispatch.loading.hide({ name, action })
```
Comment (crazy4groovy) [inline plugins/loading/src/index.ts]: omit `catch`, but decrements counter even on an error

**Comment by ShMcK** [comment]
Note that the loading plugin will likely require some changes once https://github.com/rematch/rematch/pull/236 is merged. Working on this at the moment.

At that point, loading can be handled with 2 actions, rather than the current 3.

1. show loading & action
2. hide loading

**Comment by ShMcK** [comment]
@crazy4groovy - interesting idea - an integer count is certainly more useful than a boolean. I would use this, for its debugging value alone.

The only issue I can see is that users with typechecking (TypeScript, Flow) may find it annoying. 

Do you think it's possible to make it an "opt in" feature?

**Comment by crazy4groovy** [comment]
@ShMcK interesting point about typechecking. Honestly I'm not strong on the topic so I don't have any insight, sorry. But I didn't see any `boolean` typing info that I was converting into `number` so maybe you can help me understand your concern with an example?

Boolean -> number is a breaking change, so opt-in would help maintain backwards-compatibility. That is often a smart move, unless the added feature completely trumps the existing implementation. Can you see any case where the new functionality is not better than current? Alternatively a v1.x (semver) bump may be the right move.

Thoughts?

**Comment by ShMcK** [comment]
> Can you see any case where the new functionality is not better than current?

Nope. You have a good idea.

> Help me understand your concern with an example

In JS, there is no issue with dynamically typing variables.

```js
var loading = 0
if (loading) {
  // do something
}
```

But when using TypeScript, for example, the compiler expects variables to be statically typed.

```js
var loading = 0
if (loading) {          // warning, expected a boolean
  // do something
}
```

Then you have to do some nasty coercion to make the compiler happy.

```js
var loading = 0
if (!!loading) {
  // do something
}
```

---

I believe these two solutions can co-exist very simply.

1) store loading as a number
2) if users "opt in" to a config for "valueType" = "boolean", then the number is coerced into a boolean, but defaults to "number"

What do you think?

**Comment by crazy4groovy** [comment]
> Then you have to do some nasty coercion to make the compiler happy.

I see. Yes, in JS world I would just expect to check loading like this:

```js
var loading = 0
if (loading > 0) { // always boolean
  // do something
}
```

I'm not opposed to a config param here, but it's not obvious to me how to store state as a number, and retrieve it as a boolean. Except maybe a selector of some kind? A hint here would be interesting for me.

**Comment by crazy4groovy** [comment]
^^ poke?

**Comment by ShMcK** [comment]
Sorry, I'll need a bit more time. I'll try to get back to you later this evening.

**Comment by ShMcK** [comment]
I think both solutions can co-exist. I came up with something like this:

```js
const createLoadingAction = (valueFn) =>
  (show) =>
    (state, { name, action }: any) => ({
      ...state,
      global: valueFn[show](state.global),
      models: {
        ...state.models,
        [name]: valueFn[show](state.models[name]),
      },
      effects: {
        ...state.effects,
        [name]: {
          ...state.effects[name],
          [action]: valueFn[show](state.effects[name][action]),
        },
      },
    })

 let valueFn
 let defaultValue
  if (config.valueType === 'number') {
    valueFn = {
      hide: () => 0,
      show: (v) => v + 1,
    }
   defaultValue = 0
  } else {
    valueFn = {
      hide: () => false,
      show: () => true,
    }
    defaultValue = false
  }

  const loadingAction = createLoadingAction(valueFn)

  /* add defaultValue where needed */
```

In this case, users can opt in to using numbers over booleans by adding a `config.valueType = 'number'`.

This way, TypeScript/Flow folks are happy, no breaking changes, and the README can emphasize that numbers are an excellent option. What do you think?

**Comment by crazy4groovy** [comment]
Interesting idea! Feels pretty complicated but it's an interesting concept, I'll try to take a look this weekend.

Also, I think the numerical `hide` isn't quire right, I think it should just tick down the number instead of setting it to 0.

**Comment by ShMcK** [comment]
Great, also have a look at #266. Can you think of a way to resolve this with the loading plugin?

File: plugins/loading/src/index.ts [hunk 6]
```diff
@@ -33,20 +51,32 @@ const loadingPlugin = (config: LoadingConfig = {}): any => {
...
     throw new Error('loading plugin config cannot have both a whitelist & a blacklist')
   }
+}
+
+export default (config: LoadingConfig = {}): any => {
+  validateConfig(config)
 
   const loadingModelName = config.name || 'loading'
+
+  const converter =
```
Comment (crazy4groovy) [inline plugins/loading/src/index.ts]: I'm pretty happy with this solution, it keeps a "background" counter called `cntState` of dispatched effects, but converts it onto its public state as either a number or boolean.

File: plugins/loading/test/loading-asNumber.test.js [hunk 7]
```diff
@@ -0,0 +1,327 @@
+const delay = ms => new Promise(r => setTimeout(r, ms))
```
Comment (crazy4groovy) [inline plugins/loading/test/loading-asNumber.test.js]: There are a lot more tests now.. I just duplicated them and tested for boolean in one file, and number in this one.

File: plugins/loading/README.md [hunk 8]
```diff
@@ -62,21 +62,28 @@ init({
 
 ## Options
 
-### name
+### asNumber
```
Comment (crazy4groovy) [inline plugins/loading/README.md]: yay for docs! :+1:

File: plugins/loading/examples/react-loading-example/src/index.js [hunk 9]
```diff
@@ -6,7 +6,7 @@ import createLoadingPlugin from '@rematch/loading'
 import App from './App'
 import example from './example'
 
-const loadingPlugin = createLoadingPlugin()
+const loadingPlugin = createLoadingPlugin({ asNumber: true })
```
Comment (crazy4groovy) [inline plugins/loading/examples/react-loading-example/src/index.js]: The example uses numbers, I think it's way more practical personally

**Comment by ShMcK** [comment]
Quite an elegant solution with the "converter". Thanks for putting in all the effort @crazy4groovy, great work, and sorry this took so long. I really appreciate your patience.

**Comment by ShMcK** [comment]
Getting a strange error from the Rollup bundler.

```
TypeError: parse is not a function in lib/index.js
```

WIll have to look at this tomorrow morning before I'm able to publish as `@rematch/loading@0.4`.

**Comment by crazy4groovy** [comment]
Drat sorry. Here's a 100% shot in the dark to try: could be the `export default` statement I consolidated, try pulling it back out to the bottom maybe (I've just heard someone else mention that on a different project.)

**Comment by crazy4groovy** [comment]
Maybe this issue: https://github.com/rollup/rollup-plugin-commonjs/issues/299

**Comment by ShMcK** [comment]
Something changed in Rollup. Before I could use the peer dependencies in the root "rematch" directory, but now I had to install them all as dev dependencies in the plugins.

**Comment by ShMcK** [comment]
Thanks for your help. Published as @rematch/loading@0.4.0.

**Comment by semoal** [comment]
@tianzhich

**Pull request merged**